rust/otap-dataflow/crates/otap/src/retry_processor.rs

```diff
@@ -323,208 +234,212 @@ pub fn create_retry_processor(
-                let delay_ms = (self.config.initial_retry_delay_ms as f64
-                    * self
-                        .config
-                        .backoff_multiplier
-                        .powi(pending.retry_count as i32 - 1))
-                .min(self.config.max_retry_delay_ms as f64) as u64;
-
-                pending.next_retry_time = Instant::now() + Duration::from_millis(delay_ms);
-                let retry_count = pending.retry_count;
-                let _previous = self.pending_messages.insert(id, pending);
-                if let Some(m) = self.metrics.as_mut() {
-                    m.retry_attempts.inc();
-                }
-                log::debug!("Scheduled message {id} for retry attempt {retry_count}");
-            } else {
-                log::error!(
-                    "Message {} exceeded max retries ({}), dropping. Last error: {}",
-                    id,
-                    self.config.max_retries,
-                    pending.last_error
-                );
-                if let Some(m) = self.metrics.as_mut() {
-                    m.msgs_dropped_exceeded_retries.inc();
-                }
+        effect_handler.notify_ack(ack).await?;
+
+        // Forward the ACK upstream to complete the request chain
+        match calldata.try_into() {
+            Err(_err) => {
+                // Malformed context error: we don't know what this is.
             }
-        } else {
-            log::warn!("Attempted to handle nack for non-existent message with ID: {id}");
-        }
+            Ok(RetryState { num_items, .. }) => {
+                // The producer and consumer both succeed.
+                self.metrics.add_consumed_success(signal, num_items);
+                self.metrics.add_produced_success(signal, num_items);
+            }
+        };
+
         Ok(())
     }
 
-    async fn process_pending_retries(
+    async fn handle_nack(
         &mut self,
+        mut nack: NackMsg<OtapPdata>,
         effect_handler: &mut EffectHandler<OtapPdata>,
     ) -> Result<(), Error> {
-        let now = Instant::now();
-        let mut ready_messages = Vec::new();
+        let signal = nack.refused.signal_type();
 
-        for (&id, pending) in &self.pending_messages {
-            if pending.next_retry_time <= now {
-                ready_messages.push((id, pending.data.clone()));
+        let mut rstate: RetryState = match nack.calldata.clone().try_into() {
+            Err(_err) => {
+                effect_handler.notify_nack(nack).await?;
+                // Malformed context error: we don't know what this is.
+                return Ok(());
             }
+            Ok(retry) => retry,
+        };
+
+        // Regardless of the next step, this NACK counts as a producer
+        // refusal.
+        self.metrics.add_produced_refused(signal, rstate.num_items);
+
+        // Check for missing payload, we won't retry an empty request.
+        if nack.refused.is_empty() {
```

Comment:
Do we expect NACK messages without a payload given that we subscribe to `Interests::RETURN_DATA` in the `process` method?

Reply:
Not sure if I follow -- just because something shouldn't happen won't prevent it from, in case of bugs or what not. I was checking this because we checked the same condition when the data first arrived. If the same is not still true, the payload was lost, I would say through "mis-handling". If a caller wants to use RETURN_DATA, they have to prepare for the case where the data does not return intact anyway.